### PR TITLE
[tune] implement shim instantiation

### LINKED ISSUE #10451: [tune] shim instantiation of search algorithms
### Describe your feature request

allow creating a search algorithm for `tune.run` given:
- (maybe; see option 1 below) a string defining the search algorithm to use
- (maybe; see option 2 below) the search space configuration

### API
there will be two ways to use shim instantiation (option 3 from the design doc):

#### Option 1
(notice how the search algorithm definition doesn't require redefining the search configuration when initializing `HyperOptSearch`)

> ```python
> from ray import tune
> from ray.tune.param import Float, Integer, Categorical
> from ray.tune.param import Grid, Uniform, LogUniform, Normal
> 
> space = {
>     "lr": Float(1e-4, 1e-1).LogUniform(),
>     "num_epochs": Integer(5, 10).Uniform(),
>     "temperature": Float().Normal(1, 0.1),
>     "batch_size": Categorical([32, 64]).Grid()
> }
> 
> tune.run(
>     trainable,
>     space,
>     search_alg=HyperOptSearch(metric="mean_loss"))
> ```

#### Option 2

> ```python
> from ray import tune
> from ray.tune.param import Float, Integer, Categorical
> from ray.tune.param import Grid, Uniform, LogUniform, Normal
> 
> space = {
>     "lr": Float(1e-4, 1e-1).LogUniform(),
>     "num_epochs": Integer(5, 10).Uniform(),
>     "temperature": Float().Normal(1, 0.1),
>     "batch_size": Categorical([32, 64]).Grid()
> }
> 
> search_alg = tune.create_searcher("HyperOpt", space,
>     metric="mean_loss")
> 
> tune.run(
>     trainable,
>     search_alg=search_alg)
> ```

---

related: #9969, #10401, #10444

CC @richardliaw @krfricke

### PR DESCRIPTION
**Pull request by sumanthratna**
## Why are these changes needed?

makes creating search algorithms based on a string easier (useful for CLIs where users define search alg and scheduler to use)

## Related issue number

related: #9969, #10401, #10444, #10451 (closes #10451)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

CC @richardliaw

### CODE REVIEW

**Comment by sumanthratna** [comment]
yikes this PR's commits have become way too bloated than they need to be; @richardliaw should I rebase and force-push to reduce the number of commits?

**Comment by richardliaw** [comment]
Hmm I don't really mind - up to you!

File: python/ray/tune/schedulers/__init__.py [hunk 1]
```diff
@@ -7,6 +7,64 @@
...
+def create_scheduler(
+        scheduler,
+        metric="episode_reward_mean",
+        mode="max",
+        **kwargs,
+):
+    """Instantiate a scheduler based on the given string.
+
+    This is particularly useful when you want end-users to input which
+    scheduler they want to use.
```
Comment (richardliaw) [inline python/ray/tune/schedulers/__init__.py:20]: These docstrings should be user-facing; so maybe say something like:

```
This is useful for swapping between different schedulers.
```

File: python/ray/tune/tests/test_api.py [hunk 2]
```diff
@@ -1105,6 +1107,35 @@ def train(config, reporter):
...
+                                                      **kwargs)
+        real_searcher_hyperopt = HyperOptSearch(searcher_hyperopt)
+        assert shim_searcher_hyperopt == real_searcher_hyperopt
+
+
+if __name__ == "__main__":
+    import pytest
+    import sys
+    sys.exit(pytest.main([__file__]))
+
```
Comment (richardliaw) [inline python/ray/tune/tests/test_api.py:1138]: duplicate of below

File: python/ray/tune/suggest/__init__.py [hunk 3]
```diff
@@ -5,6 +5,88 @@
...
+        "skopt": _import_skopt_search,
+        "hyperopt": _import_hyperopt_search,
+        "bayesopt": _import_bayesopt_search,
+        "bohb": _import_bohb_search,
+        "nevergrad": _import_nevergrad_search,
+        "optuna": _import_optuna_search,
+        "zoopt": _import_zoopt_search,
+        "sigopt": _import_sigopt_search,
+    }
+    SearcherClass = SEARCH_ALG_IMPORT[search_alg]()
```
Comment (richardliaw) [inline python/ray/tune/suggest/__init__.py:86]: Nice! 
Here, can you do `search_alg.lower()` and also provide a nice validation error message?

```
search_alg = search_alg.lower()
if search_alg not in SEARCH_ALG_IMPORT:
    raise ValueError(f"Search alg must be one of {list(SEARCH_ALG_IMPORT)}. Got: {search_alg}")
```
Same with schedulers

**Review by richardliaw — COMMENTED** [review]
Looks good to me, pending 1 change.

**Review by richardliaw — APPROVED** [review]
Looks great! Do you mind adding this into the docs too? `ray/docs/source/tune/api_docs/{suggestion / schedulers}.rst`

**Comment by richardliaw** [comment]
bump - looks like tests are failing. I think you should do `type(scheduler) == type(...)` in the unit tests

**Comment by sumanthratna** [comment]
> Looks great! Do you mind adding this into the docs too? `ray/docs/source/tune/api_docs/{suggestion / schedulers}.rst`

sure, I'll try to do that soon

> looks like tests are failing. I think you should do `type(scheduler) == type(...)` in the unit tests

Oh—I had (lazily) assumed that schedulers overrode the equality operator. Is this something we should implement? I don't think it would be too hard (untested):
```python
def __eq__(self, other):
    if type(self) != type(other):
        return False
    return self.__dict__ == other.__dict__
```

this is probably better suited for a different PR, though

**Comment by richardliaw** [comment]
Hm I guess I don't know if those semantics really are needed since you're almost certainly only using 1 scheduler.

**Comment by richardliaw** [comment]
Thanks for putting together this PR though! Definitely looking forward to getting this merged :)

**Comment by sumanthratna** [comment]
@richardliaw can you take a look at 6dcf1f0b9387b9f02e5d26a113a52c7b3fcea2a7 and let me know if I should change anything else? once you take a look I'll add to the schedulers doc too

**Comment by richardliaw** [comment]
looks good!

**Comment by richardliaw** [comment]
Hey @sumanthratna looks like the create search alg test is still broken

Don't forget to run `ci/format.sh` too to address lint test!

**Comment by sumanthratna** [comment]
> Hey @sumanthratna looks like the create search alg test is still broken

@richardliaw ugh I had copied the example from the docs because I wanted to write the test correctly but I was looking at the 0.8.7 docs instead of master...fix is (hopefully) on the way

**Comment by richardliaw** [comment]
hmm tests still failing?

**Comment by sumanthratna** [comment]
we can wait until all tests finish but @richardliaw I think this is finally ready for merging
